File: codigo/backend/go-backend/routes/usuario_routes.go
package routes

import (
    "net/http"
    "go-backend/models"
    "go-backend/db"
    "github.com/gin-gonic/gin"
)

func RegisterUsuarioRoutes(router *gin.Engine) {
    router.GET("/usuarios", getUsuarios)
    router.GET("/usuarios/:id", getUsuario)
    router.POST("/usuarios", createUsuario)
    router.PATCH("/usuarios/:id", updateUsuario)
    router.DELETE("/usuarios/:id", deleteUsuario)
}

func getUsuarios(c *gin.Context) {
    var usuarios []models.Usuario
    if err := db.DB.Find(&usuarios).Error; err != nil {
        c.JSON(http.StatusInternalServerError, gin.H{"message": "Erro ao buscar usuários"})
        return
    }
    c.JSON(http.StatusOK, usuariosToMap(usuarios))
}

func getUsuario(c *gin.Context) {
    var usuario models.Usuario
    id := c.Param("id")
    if err := db.DB.First(&usuario, id).Error; err != nil {
        c.JSON(http.StatusNotFound, gin.H{"message": "Usuário não encontrado"})
        return
    }
    c.JSON(http.StatusOK, usuario.ToMap())
}

func createUsuario(c *gin.Context) {
    var usuario models.Usuario
    if err := c.ShouldBindJSON(&usuario); err != nil {
        c.JSON(http.StatusBadRequest, gin.H{"message": "Dados inválidos"})
        return
    }
    if err := db.DB.Create(&usuario).Error; err != nil {
        c.JSON(http.StatusInternalServerError, gin.H{"message": "Erro ao criar usuário"})
        return
    }
    c.JSON(http.StatusCreated, usuario.ToMap())
}

func updateUsuario(c *gin.Context) {
    var usuario models.Usuario
    id := c.Param("id")
    if err := db.DB.First(&usuario, id).Error; err != nil {
        c.JSON(http.StatusNotFound, gin.H{"message": "Usuário não encontrado"})
        return
    }

    var data map[string]interface{}
    if err := c.ShouldBindJSON(&data); err != nil {
        c.JSON(http.StatusBadRequest, gin.H{"message": "Dados inválidos"})
        return
    }

    if err := db.DB.Model(&usuario).Updates(data).Error; err != nil {
        c.JSON(http.StatusInternalServerError, gin.H{"message": "Erro ao atualizar usuário"})
        return
    }

    c.JSON(http.StatusOK, usuario.ToMap())
}

func deleteUsuario(c *gin.Context) {
    var usuario models.Usuario
    id := c.Param("id")
    if err := db.DB.First(&usuario, id).Error; err != nil {
        c.JSON(http.StatusNotFound, gin.H{"message": "Usuário não encontrado"})
        return
    }

    if err := db.DB.Delete(&usuario).Error; err != nil {
        c.JSON(http.StatusInternalServerError, gin.H{"message": "Erro ao excluir usuário"})
        return
    }

    c.JSON(http.StatusOK, gin.H{"message": "Usuário excluído"})
}

func usuariosToMap(usuarios []models.Usuario) []map[string]interface{} {
    result := make([]map[string]interface{}, len(usuarios))
    for i, usuario := range usuarios {
        result[i] = usuario.ToMap()
    }
    return result
}

File: codigo/backend/sap-ecc-emulator/src/routes/usuario_routes.py
from flask import Blueprint, jsonify, request, make_response
from models.usuario import Usuario
from db import db

usuario_bp = Blueprint('usuario_bp', __name__)

@usuario_bp.route('/usuarios', methods=['GET'])
def get_usuarios():
    usuarios = Usuario.query.all()
    return jsonify([usuario.to_dict() for usuario in usuarios])

@usuario_bp.route('/usuarios/<int:id>', methods=['GET'])
def get_usuario(id):
    usuario = Usuario.query.get(id)
    if usuario:
        return jsonify(usuario.to_dict())
    else:
        return make_response(jsonify(message="Usuário não encontrado"), 404)

@usuario_bp.route('/usuarios', methods=['POST'])
def create_usuario():
    data = request.json
    usuario = Usuario(**data)
    db.session.add(usuario)
    db.session.commit()
    return jsonify(usuario.to_dict()), 201

@usuario_bp.route('/usuarios/<int:id>', methods=['PATCH'])
def update_usuario(id):
    usuario = Usuario.query.get(id)
    if usuario:
        data = request.json
        for key, value in data.items():
            setattr(usuario, key, value)
        db.session.commit()
        return jsonify(usuario.to_dict())
    else:
        return make_response(jsonify(message="Usuário não encontrado"), 404)

@usuario_bp.route('/usuarios/<int:id>', methods=['DELETE'])
def delete_usuario(id):
    usuario = Usuario.query.get(id)
    if usuario:
        db.session.delete(usuario)
        db.session.commit()
        return jsonify(message="Usuário excluído")
    else:
        return make_response(jsonify(message="Usuário não encontrado"), 404)
